src/main.rs

```diff
@@ -30,82 +31,126 @@ mod error;
     println!("Sending recorded events: {} {}",
              out_info.id(),
              out_info.name());
-    let out_port = context.output_port(out_info, 1024).unwrap();
+    let out_port = try!(context.output_port(out_info, 1024));
+    let looper = looper::Looper::new(PortMidiNoteTracker::new(out_port));
+    Ok(looper)
+}
 
-    let window_width = RATIO_WIDTH * RATIO_FACTOR;
+fn create_event_loop(context: &pm::PortMidi, input_id: DeviceId) -> Result<EventLoop<'static>> {
+    let window_width = RATIO_WIDTH  * RATIO_FACTOR;
     let window_height = RATIO_HEIGHT * RATIO_FACTOR;
-    let sdl_context = sdl2::init().unwrap();
-    let video_subsystem = sdl_context.video().unwrap();
-    let timer_subsystem = sdl_context.timer().unwrap();
-    let ttf_context = sdl2_ttf::init().unwrap();
 
-    let window = video_subsystem.window("Dimooper", window_width, window_height)
+    let in_info = try!(context.device(input_id));
+    println!("Listening on: {} {}", in_info.id(), in_info.name());
+    let in_port = try!(context.input_port(in_info, 1024));
+
+    let sdl_context = try!(sdl2::init());
+    let video_subsystem = try!(sdl_context.video());
+    let timer_subsystem = try!(sdl_context.timer());
+
+    let window = try!(video_subsystem.window("Dimooper", window_width, window_height)
         .position_centered()
         .opengl()
-        .build()
-        .unwrap();
-
-    let renderer = window.renderer().build().unwrap();
+        .build());
 
-    let bpm_popup = {
-        let font = ttf_context.load_font(Path::new(TTF_FONT_PATH), 50).unwrap();
-        Popup::new(font)
-    };
+    let renderer = try!(window.renderer().build());
+    let event_pump = try!(sdl_context.event_pump());
+    let event_loop = EventLoop::new(timer_subsystem, event_pump, in_port, renderer);
+    Ok(event_loop)
+}
 
-    let event_pump = sdl_context.event_pump().unwrap();
+fn init_font() -> Result<Popup> {
+    let ttf_context = try!(sdl2_ttf::init());
+    let font = try!(ttf_context.load_font(Path::new(TTF_FONT_PATH), 50));
+    let popup = Popup::new(font);
+    Ok(popup)
+}
 
-    let looper = looper::Looper::new(PortMidiNoteTracker::new(out_port));
+fn main() {
+    use clap::{App, Arg, ArgMatches};
+
+    let context = pm::PortMidi::new().expect("Unable to initialize PortMidi");
+    let devices = context.devices()
+        .expect("Unable to get list of devices")
+        .iter()
+        .map(ToString::to_string)
+        .collect::<Vec<_>>()
+        .join("\n");
+
+    let matches = App::new("Dimooper")
+        .about("Digital music looper")
+        .after_help(format!("Avaliable devices:\n{}", devices).as_ref())
+        .arg(Arg::with_name("INPUT_ID")
+            .help("Input device id")
+            .index(1)
+            .required(true))
+        .arg(Arg::with_name("OUTPUT_ID")
+            .help("Output device id")
+            .index(2)
+            .required(true))
+        .arg(Arg::with_name("SCREEN")
```

Comment:
> This will lead to some code duplication as both main and keyboard subcommands require same `input_id` and `output_id` arguments.

YES! That was exactly my struggle on the stream! :D Honestly, I don't know why I always overreact over such stupid things as code duplicate, I'm really sorry for my behaviour. :) Probably bad tea, I dunno...

I still insist on having the subcommand approach. I'm ok with the lack of the default command. Let's make it like:

```console
$ dimooper looper 1 2    # looper mode
$ dimooper keyboard 1 2  # keyboard configuration mode
```

I'm also ok with the code duplicate for now. I think later when we have more modes it's gonna be more clear how to generalize all of that.